src/lib/core/fiber_channel.c

```diff
@@ -191,7 +205,24 @@ fiber_channel_check_wait(struct fiber_channel *ch, ev_tstamp start_time,
 void
 fiber_channel_close(struct fiber_channel *ch)
 {
-	if (ch->is_closed)
+	if (!fiber_channel_graceful_close || ch->count == 0) {
+		/* Destroy if no graceful close or buffer is already empty. */
+		fiber_channel_destroy(ch);
+		return;
+	}
+	ch->is_closed = true;
+	/* Wake up all channel_put(), they will automatically cancel. */
+	struct fiber *f;
+	while (!rlist_empty(&ch->waiters)) {
+		f = rlist_first_entry(&ch->waiters, struct fiber, state);
+		fiber_channel_waiter_wakeup(f, FIBER_CHANNEL_WAIT_CLOSED);
+	}
```

Comment:
AFAIU a channel with an empty buffer has either readers or writers, because a writer removes a reader from the list and vice versa (see comments in `fiber_channel_waiter_wakeup()`).

While thinking on it I found that we have no test case that gracefully close a channel with a waiting reader. It'll have no buffered messages, but anyway. Let's add a unit test case and a Lua test case.

Reply:
Igor added such test cases.